api_server.go

```diff
@@ -68,39 +70,48 @@ func logRequest(handlerFunc APIHandlerFunc, r *http.Request) ([]byte, int) {
 	case code >= 500:
 		logger.WithFields(fields).Error("request failed")
 	case 400 <= code && code < 500:
-		logger.WithFields(fields).Error("request failed")
+		logger.WithFields(fields).Error("invalid request")
 	default:
 		logger.WithFields(fields).Info("request completed")
 	}
 	return body, code
 }
 
-func (x *ApiServer) MusicPDFsIndex(r *http.Request) ([]byte, int) {
+func (x *ApiServer) SheetsIndex(r *http.Request) ([]byte, int) {
 	// only accept get
 	if r.Method != http.MethodGet {
 		return nil, http.StatusMethodNotAllowed
 	}
 
-	objects := x.ListObjects(MusicPdfsBucketName)
-	pdfs := make([]MusicPDFMeta, 0, len(objects))
+	type tableRow struct {
+		Sheet
+		Link string `json:"link"`
+	}
+
+	objects := x.ListObjects(SheetsBucketName)
+	rows := make([]tableRow, 0, len(objects))
 	for i := range objects {
-		pdfs = append(pdfs, NewMusicPDFMetaFromTags(objects[i].Tags))
+		sheet := NewSheetFromTags(objects[i].Tags)
+		rows = append(rows, tableRow{
+			Sheet: sheet,
+			Link:  sheet.Link(),
+		})
 	}
 
 	var buf bytes.Buffer
 	switch true {
 	case acceptsType(r, "text/html"):
-		musicPDFsTemplate, err := template.ParseFiles("public/music_pdfs.gohtml")
+		sheetsTemplate, err := template.ParseFiles("public/sheets.gohtml")
```

Comment:
IIRC we can just parse the template once, and execute it on fresh output buffers.

Reply:
Yes, we should do that later. It can parse all the templates at startup and serve them from mem. However, it makes developing the templates a pain because I would have to restart the app every time :sweat: